adr/0003-jsonpath-support.md

```diff
@@ -0,0 +1,78 @@
+# 3. JSONPath support
+
+Date: 15-03-2024
+
+## Status
+
+Draft
+
+## Context
+
+A json may contain multiple nested objects with name/value pairs that have the same name but require a different masking strategy.  
+For example, in the following json
+
+```json
+{
+  "name": "do not mask",
+  "nestedObject": {
+    "name": "mask",
+    "public": "do not mask",
+    "private": "mask"
+  }
+}
+```
+The outer `name` value must not be masked but the inner `name` value in `nestedObject` must be masked.  
+This cannot be achieved with regular key masking.
+
+## Decisions
+
+### JsonPATH  
+
+The solution to the described problem is to provide a way to disambiguate name/value pairs by selecting the target pair using its location in json.  
+The industry standard for selecting values in json is JsonPATH. Most developers are expected to know how to create basic JsonPATH queries using either bracket or dot notation.  
+Therefore, we decided to solve the name ambiguity problem using JsonPATH.  
+
+### Supported features  
+
+JsonPATH RFC specifies a wide variety of features. Not all of them are required to solve the described problem.  
+Therefore, we have to decide which features are necessary and which are not. 
+  
+The decision is premised on the following expectations:
+1. Most of the clients will not face the name ambiguity problem, therefore they will not use JsonPATH.  
+2. Correctness, performance and maintainability of the library takes precedent over the amount of features.  
+3. JsonPATH support is required only for solving the name ambiguity problem.  
+
+Any implemented JsonPATH feature adds some performance overhead both on configuration and on masking time, even if it is not used. In general, it also increases the likelihood of incorrect behaviour as the amount of code increases. Therefore, having unnecessary features violates the premises.   
+Therefore, we decided to keep support for JsonPATH as limited as possible and extend it only when someone specifically requests it.
+
+The following features are necessary to support:
+1. **Support for bracket, dot and mixed notations**.   
+We expect developers to know either of these notations, but we cannot know which one exactly.  
+2. **Support for segments**.  
+This is the main building block of JsonPATH queries.
+3. **Support for wildcard segments and selectors**.  
+Wildcard selectors are necessary for traversing array values. 
+
+## Not supported features
+
+1. **Name selectors**.  
+Using segments is enough to select any value in the json.  
```

Comment:
I thought we fully support name selectors, if I understand it correctly, in the spec "segment" is used to describe any segment, including name selector  which is the simplest selector (`$.a` is a name selector)

Reply:
I think you are right. Let me fix that.